client/lib/plugins/sanitize-section-content.js

```diff
@@ -0,0 +1,190 @@
+		}
+	}
+
+	if ( 'img' === tagName && ( 'alt' === attrName || 'src' === attrName ) ) {
+		return true;
+	}
+
+	return false;
+};
+
+const isValidYoutubeEmbed = node => {
+	if ( node.nodeName.toLowerCase() !== 'iframe' ) {
+		return false;
+	}
+
+	if ( node.getAttribute( 'class' ) !== 'youtube-player' ) {
+		return false;
+	}
+
+	if ( node.getAttribute( 'type' ) !== 'text/html' ) {
+		return false;
+	}
+
+	const link = document.createElement( 'a' );
+	link.href = node.getAttribute( 'src' );
+
+	return (
+		validUrl.isWebUri( node.getAttribute( 'src' ) ) &&
+		( link.hostname === 'youtube.com' || link.hostname === 'www.youtube.com' )
+	);
+};
+
+/**
+ * Sanitizes input HTML for security and styling
+ *
+ * @param {String} content unverified HTML
+ * @returns {string} sanitized HTML
+ */
+export const sanitizeSectionContent = content => {
+	const parser = new DOMParser();
+	const doc = parser.parseFromString( content, 'text/html' );
+
+	// this will let us visit every single DOM node programmatically
+	const walker = doc.createTreeWalker( doc.body );
+
+	// we don't want to remove nodes while walking the tree
+	// or we'll invite data-race bugs. instead, we'll track
+	// which ones we want to remove then drop them at the end
+	const removeList = [];
+
+	// walk over every DOM node
+	while ( walker.nextNode() ) {
+		const node = walker.currentNode;
+		const tagName = node.nodeName.toLowerCase();
+
+		if ( ! isAllowedTag( tagName ) && ! isValidYoutubeEmbed( node ) ) {
+			removeList.push( node );
+			continue;
+		}
+
+		// strip out anything not explicitly allowed
+		// in the attributes. we want to eliminate
+		// potential cross-site scripting attacks _and_
+		// prevent custom styles from interfering with
+		// our page's own rendering
+
+		const attrRemoveList = [];
+
+		// these aren't Arrays, so iteration is odd
+		for ( let i = 0; node.attributes && i < node.attributes.length; i++ ) {
+			const { name: attrName, value } = node.attributes[ i ];
+
+			// get rid of non-whitelisted attributes
+			if ( ! isAllowedAttr( tagName, attrName ) ) {
+				attrRemoveList.push( attrName );
+				continue;
+			}
+
+			// for some, specifically links, we don't want to
+			// "sanitize" the values because we could mess them up
+			// for example, encoding a real URL will break it
+			if (
+				( ( 'src' === attrName || 'href' === attrName ) && validUrl.isWebUri( value ) ) ||
+				( 'iframe' === tagName && 'type' === attrName && isValidYoutubeEmbed( node ) )
+			) {
+				continue;
+			}
+
+			// just don't even allow non-url links
+			if ( ( 'href' === attrName || 'src' === attrName ) && ! validUrl.isWebUri( value ) ) {
+				attrRemoveList.push( attrName );
+				continue;
+			}
+
+			// every other other allowed attribute must be sanitized
+			// this is a broad and clumsy stroke but it should be
+			// effective, leaning towards safety instead of
+			// compatibility with the remote HTML
+			node.setAttribute( attrName, encodeURIComponent( value ) );
```

Comment:
Why do we URL encode attributes?

Reply:
It was my attempt at a blanket protection while still allowing some resemblance of the original attributes through. The alternative here is probably just eliminating all attributes

Reply:
It seems to me that this encoding is not protecting against any actual threats. `isAllowedAttr` already removes almost all attributes. The remaining few are either `href` or `src` that are already sanitized by `isWebUri`, or are never interpreted as URL or script (`alt`, `class`, ...)

And it does harm to `<img alt>` attributes by mangling them from `"alt text?"` to `"alt%20text%3F"`. The encoded form will be rendered.

While testing this, I got an idea that it could be useful to allow a `title` attribute on images and other elements? It gets displayed as tooltip on hover. `alt` is displayed only if the image fails to load.

Reply:
You're right in that none of the attributes we allow would open up attack threats. I guess this would be there more for cases where we have links or event handlers, but links already need to be `http` or `https` URLs.

I'll just back this out; I think that's legitimate